substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1040 @@
+		StorageMap<_, Identity, u32, BoundedVec<TicketId, ConstU32<SEGMENT_MAX_SIZE>>, ValueQuery>;
+
+	/// Parameters used to construct the epoch's ring verifier.
+	///
+	/// In practice: Updatable Universal Reference String and the seed.
+	#[pallet::storage]
+	#[pallet::getter(fn ring_context)]
+	pub type RingContext<T: Config> = StorageValue<_, vrf::RingContext>;
+
+	/// Ring verifier data for the current epoch.
+	#[pallet::storage]
+	pub type RingVerifierData<T: Config> = StorageValue<_, vrf::RingVerifierData>;
+
+	/// Slot claim vrf-preoutput used to generate per-slot randomness.
+	///
+	/// The value is ephemeral and is cleared on block finalization.
+	#[pallet::storage]
+	pub(crate) type ClaimTemporaryData<T> = StorageValue<_, vrf::VrfOutput>;
+
+	/// Genesis configuration for Sassafras protocol.
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		/// Genesis authorities.
+		pub authorities: Vec<AuthorityId>,
+		/// Genesis epoch configuration.
+		pub epoch_config: EpochConfiguration,
+		/// Phantom config
+		#[serde(skip)]
+		pub _phantom: sp_std::marker::PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			EpochConfig::<T>::put(self.epoch_config);
+			Pallet::<T>::initialize_genesis_authorities(&self.authorities);
+
+			#[cfg(feature = "construct-dummy-ring-context")]
+			{
+				debug!(target: LOG_TARGET, "Constructing dummy ring context");
+				let ring_ctx = vrf::RingContext::new_testing();
+				RingContext::<T>::put(ring_ctx);
+				Pallet::<T>::update_ring_verifier(&self.authorities);
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Block initialization.
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			let claim = <frame_system::Pallet<T>>::digest()
+				.logs
+				.iter()
+				.filter_map(|item| item.pre_runtime_try_to::<SlotClaim>(&SASSAFRAS_ENGINE_ID))
+				.next()
+				.expect("Valid block must have a slot claim. qed");
+
+			CurrentSlot::<T>::put(claim.slot);
+
+			// As the slots may not be zero-based, we need to keep track of what is the
+			// slot used fo the first block.
+			if <frame_system::Pallet<T>>::block_number() == One::one() {
+				GenesisSlot::<T>::put(claim.slot);
+
+				// Deposit a log as this is the first block in first epoch.
+				let next_epoch = NextEpochDescriptor {
+					randomness: Self::next_randomness(),
+					authorities: Self::next_authorities().into_inner(),
+					config: None,
+				};
+				Self::deposit_next_epoch_descriptor_digest(next_epoch);
+			}
+
+			let randomness_output = claim
+				.vrf_signature
+				.outputs
+				.get(0)
+				.expect("Presence should have been already checked by the client; qed");
+			ClaimTemporaryData::<T>::put(randomness_output);
+
+			// Enact epoch change, if necessary.
+			T::EpochChangeTrigger::trigger::<T>(now);
+
+			Weight::zero()
```

Comment:
This has been zero since the begin, when I basically started from the Babe code.

https://github.com/paritytech/polkadot-sdk/blob/8d2637905ba920dd1f0e8f1212ec98e45420f514/substrate/frame/babe/src/lib.rs#L345-L348

Now I updated it by getting the weight from a new benchmark.

The weight is now the sum of the `on_initialize` weight plus the weight returned by the `EpochChangeTrigger::trigger` hook.

This is important, as it may trigger the recomputation of the Ring verification key, which has a non-negligible cost

Reply:
https://github.com/paritytech/polkadot-sdk/pull/1577/commits/d9607b6c8ed29c15139cb4a15cf3eed709bcbed9